org.eclipse.lsp4e/src/org/eclipse/lsp4e/LanguageClientImpl.java

```diff
@@ -83,6 +94,51 @@ public final void logMessage(MessageParams message) {
 		CompletableFuture.runAsync(() -> ServerMessageHandler.logMessage(wrapper, message));
 	}
 
+	private Map<String, BlockingQueue<ProgressParams>> progressQueueMap = new HashMap<>();
+
+	@Override
+	public CompletableFuture<Void> createProgress(WorkDoneProgressCreateParams params) {
+		String jobName = params.getToken().getLeft();
+		LinkedBlockingDeque<ProgressParams> blockingQueue = new LinkedBlockingDeque<ProgressParams>();
+		progressQueueMap.put(jobName, blockingQueue);
+
+		Job job = Job.create(jobName, new ICoreRunnable() {
+			@Override
+			public void run(IProgressMonitor monitor) throws CoreException {
+				ProgressParams nextNotification;
+				try {
+					while ((nextNotification = blockingQueue.take()) != null) {
+						WorkDoneProgressNotification nextProgressNotification = nextNotification.getValue().getLeft();
+						if (nextProgressNotification != null) {
+							WorkDoneProgressKind kind = nextProgressNotification.getKind();
+							if (kind == WorkDoneProgressKind.begin) {
+								monitor.beginTask(jobName, 100);
+							} else if (kind == WorkDoneProgressKind.report) {
+								WorkDoneProgressReport report = (WorkDoneProgressReport) nextProgressNotification;
+								monitor.worked(report.getPercentage());
```

Comment:
This doesn't take into account that in LSP, a percentage is reported, a progress monitor is accumulating progress. So the reported percentage should ideally take the already reported progress (how many of 100) into account and only add the delta to the progress monitor.

Reply:
Thanks for the hint.